edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,693 @@
+    using Microsoft.Extensions.Logging;
+
+    public class RegistryController : Controller
+    {
+        const string ContentType = "application/json; charset=utf-8";
+
+        readonly IRegistryOnBehalfOfApiClient apiClient;
+        readonly Task<IHttpRequestAuthenticator> authenticatorGetter;
+        readonly Task<IEdgeHub> edgeHubGetter;
+
+        public RegistryController(
+            IRegistryOnBehalfOfApiClient apiClient,
+            Task<IEdgeHub> edgeHubGetter,
+            Task<IHttpRequestAuthenticator> authenticatorGetter)
+        {
+            this.apiClient = Preconditions.CheckNotNull(apiClient, nameof(apiClient));
+            this.authenticatorGetter = Preconditions.CheckNotNull(authenticatorGetter, nameof(authenticatorGetter));
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHubGetter, nameof(edgeHubGetter));
+        }
+
+        [HttpPut]
+        [Route("devices/{deviceId}/modules/{moduleId}")]
+        public async Task CreateOrUpdateModuleAsync(
+            [FromRoute] string deviceId,
+            [FromRoute] string moduleId,
+            [FromHeader(Name="if-match")] string ifMatchHeader,
+            [FromBody] Module module)
+        {
+            try
+            {
+                Events.ReceivedRequest(nameof(this.CreateOrUpdateModuleAsync), deviceId, moduleId);
+
+                try
+                {
+                    deviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId)));
+                    moduleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(moduleId, nameof(moduleId)));
+                    Preconditions.CheckNotNull(module, nameof(module));
+
+                    if (!string.Equals(deviceId, module.DeviceId) || !string.Equals(moduleId, module.Id))
+                    {
+                        throw new ApplicationException("Device Id or module Id doesn't match between request URI and body.");
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(nameof(this.CreateOrUpdateModuleAsync), ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, FormatErrorResponseMessage(ex.Message));
+                    return;
+                }
+
+                if (!await this.AuthenticateAsync(deviceId, Option.None<string>()))
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                string targetId = $"{deviceId}/{moduleId}";
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+
+                (bool authorized, string authChain) = await this.AuthorizeActorAsync(identitiesCache, edgeDeviceId, Constants.EdgeHubModuleId, targetId);
```

Comment:
>AuthorizeActorAsync [](start = 65, length = 19)

not necessary for existing API calls, but fine for now as I need to come back and remove these anyways.  All AuthZ checks will be handled by the DeviceScopeAuthenticator as of my pending change.